File: src/Algorithms/Sorting/Selection Sort/code.py
from AlgorithmVisualizer import Tracer, Array1DTracer, LogTracer, Layout, VerticalLayout, ChartTracer, randomize

# define tracer variables {
chart = ChartTracer("Selection Sort - Graph")
tracer = Array1DTracer("Selection Sort - Array")
logger = LogTracer()
Layout.setRoot(VerticalLayout([chart, tracer, logger]))
array = randomize.Array1D(15).array
tracer.set(array)
tracer.chart(chart)
Tracer.delay()
# }

string_array = [str(int) for int in array]
println('original array = ['+','.join(string_array)+']')

for i in range(len(array)):
    min = i
    # visualize {
    tracer.select(i)
    Tracer.delay()
    # }
    for j in range(i+1, len(array)):
        # visualize {
        tracer.select(j)
        Tracer.delay()
        # }
        if array[min] > array[j]:
            tracer.depatch(min)
            min = j
            # visualize {
            tracer.patch(j)
            Tracer.delay()
            # }
        # visualize {
        tracer.deselect(j)
        # }
    println('swap '+str(array[i])+' and '+str(array[min]))
    array[min], array[i] = array[i], array[min]
    # visualize {
    tracer.patch(i, array[i])
    tracer.patch(min, array[min])
    Tracer.delay()
    tracer.depatch(i)
    # }
    # visualize {
    tracer.depatch(min)
    tracer.deselect(i)
    tracer.selectTrue(i)
    # }

string_array = [str(int) for int in array]
println('sorted array = ['+','.join(string_array)+']')

File: src/Algorithms/Sorting/Selection Sort/code.js
// import visualization libraries {
const {
  Tracer,
  Array1DTracer,
  ChartTracer,
  LogTracer,
  Randomize,
  Layout,
  VerticalLayout,
} = require("algorithm-visualizer");
// }

// define tracer variables {
const chart = new ChartTracer("Selection Sort - Chart ");
const tracer = new Array1DTracer("Selection Sort - Array ");
const logger = new LogTracer();
Layout.setRoot(new VerticalLayout([chart, tracer, logger]));
const D = Randomize.Array1D({ N: 15 });
tracer.set(D);
tracer.chart(chart);
Tracer.delay();
// }

// logger {
logger.println(`original array = [${D.join(", ")}]`);
// }
for (let i = 0; i < D.length - 1; i++) {
  let minJ = i;
  // visualize {
  tracer.select(i);
  Tracer.delay();
  // }
  for (let j = i + 1; j < D.length; j++) {
    // visualize {
    tracer.select(j);
    Tracer.delay();
    // }
    if (D[j] < D[minJ]) {
      tracer.depatch(minJ);
      minJ = j;
      // visualize {
      tracer.patch(j);
      Tracer.delay();
      // }
    }
    // visualize {
    tracer.deselect(j);
    // }
  }
  if (minJ !== i) {
    // logger {
    logger.println(`swap ${D[i]} and ${D[minJ]}`);
    // }
    const temp = D[i];
    D[i] = D[minJ];
    D[minJ] = temp;
    // visualize {
    tracer.patch(i, D[i]);
    tracer.patch(minJ, D[minJ]);
    Tracer.delay();
    tracer.depatch(i);
    // }
  }
  // visualize {
  tracer.depatch(minJ);
  tracer.deselect(i);
  tracer.selectTrue(i);
  // }
}
// logger {
tracer.selectTrue(D.length - 1);
// }
logger.println(`sorted array = [${D.join(", ")}]`);
